### PR TITLE
gh-105189: fix importlib.resources.abc deprecation docs

### PR DESCRIPTION
**Pull request by graingert**
* Issue: gh-105189

----
:books: Documentation preview :books:: https://cpython-previews--105232.org.readthedocs.build/

### CODE REVIEW

**Comment by graingert** [comment]
what to do about importlib.resources.abc.ResourceReader it seems it's supposed to be deprecated, but there's no deprecation warning in the source code

**Comment by graingert** [comment]
preview is here: https://cpython-previews--105232.org.readthedocs.build/en/105232/library/importlib.html#importlib.abc.ResourceReader

**Comment by graingert** [comment]
the versionadded entries are incorrect on the 3.11 doc pages https://docs.python.org/3.11/library/importlib.resources.abc.html#importlib.resources.abc.TraversableResources

File: Doc/library/importlib.resources.abc.rst [hunk 1]
```diff
@@ -43,7 +43,7 @@
     :const:`None`. An object compatible with this ABC should only be
     returned when the specified module is a package.
 
-    .. versionadded:: 3.7
+    .. versionadded:: 3.11
```
Comment (graingert) [inline Doc/library/importlib.resources.abc.rst]: probably don't need these as the module has `.. versionadded:: 3.11`
```suggestion
```

File: Doc/library/importlib.resources.abc.rst [hunk 2]
```diff
@@ -95,10 +95,7 @@
     For a representation of the object on the file-system, use
     :meth:`importlib.resources.as_file`.
 
-    .. versionadded:: 3.9
-
-    .. deprecated-removed:: 3.12 3.14
-       Use :class:`importlib.resources.abc.Traversable` instead.
+    .. versionadded:: 3.11
```
Comment (graingert) [inline Doc/library/importlib.resources.abc.rst]:
```suggestion
```

File: Doc/library/importlib.resources.abc.rst [hunk 3]
```diff
@@ -153,10 +150,7 @@
     Loaders that wish to support resource reading are expected to
     implement this interface.
 
-    .. versionadded:: 3.9
-
-    .. deprecated-removed:: 3.12 3.14
-       Use :class:`importlib.resources.abc.TraversableResources` instead.
+    .. versionadded:: 3.11
```
Comment (graingert) [inline Doc/library/importlib.resources.abc.rst]:
```suggestion
```

File: Doc/library/importlib.resources.abc.rst (hunk 1 above)
Comment (graingert) [reply Doc/library/importlib.resources.abc.rst]: maybe I should fix this in another PR so it can be backported to 3.11?

Comment (hugovk) [reply Doc/library/importlib.resources.abc.rst]: Sure, or make a new PR into 3.12 and backport that 3.11?

File: Doc/library/importlib.rst [hunk 4]
```diff
@@ -645,6 +645,160 @@ ABC hierarchy::
...
+
+       Abstract. The base name of this object without any parent references.
+
+    .. abstractmethod:: iterdir()
+
+       Yield Traversable objects in self.
+
+    .. abstractmethod:: is_dir()
+
+       Return True if self is a directory.
```
Comment (hugovk) [inline Doc/library/importlib.rst]:
```suggestion
       Return ``True`` if ``self`` is a directory.
```

File: Doc/library/importlib.rst [hunk 5]
```diff
@@ -645,6 +645,160 @@ ABC hierarchy::
...
+
+    .. abstractmethod:: is_file()
+
+       Return True if self is a file.
```
Comment (hugovk) [inline Doc/library/importlib.rst]:
```suggestion
       Return ``True`` if ``self`` is a file.
```

File: Doc/library/importlib.rst [hunk 6]
```diff
@@ -645,6 +645,160 @@ ABC hierarchy::
...
+    .. deprecated-removed:: 3.12 3.14
+       Use :class:`importlib.resources.abc.Traversable` instead.
+
+    .. attribute:: name
+
+       Abstract. The base name of this object without any parent references.
+
+    .. abstractmethod:: iterdir()
+
+       Yield Traversable objects in self.
```
Comment (hugovk) [inline Doc/library/importlib.rst]:
```suggestion
       Yield ``Traversable`` objects in ``self``.
```

File: Doc/library/importlib.rst [hunk 7]
```diff
@@ -645,6 +645,160 @@ ABC hierarchy::
...
+
+       Return True if self is a directory.
+
+    .. abstractmethod:: is_file()
+
+       Return True if self is a file.
+
+    .. abstractmethod:: joinpath(child)
+
+       Return Traversable child in self.
```
Comment (hugovk) [inline Doc/library/importlib.rst]:
```suggestion
       Return Traversable child in ``self``.
```

File: Doc/library/importlib.rst [hunk 8]
```diff
@@ -645,6 +645,160 @@ ABC hierarchy::
...
+
+    .. abstractmethod:: __truediv__(child)
+
+       Return Traversable child in self.
```
Comment (hugovk) [inline Doc/library/importlib.rst]:
```suggestion
       Return ``Traversable`` child in ``self``.
```

File: Doc/library/importlib.rst [hunk 9]
```diff
@@ -645,6 +645,160 @@ ABC hierarchy::
...
+
+       *mode* may be 'r' or 'rb' to open as text or binary. Return a handle
+       suitable for reading (same as :attr:`pathlib.Path.open`).
+
+       When opening as text, accepts encoding parameters such as those
+       accepted by :attr:`io.TextIOWrapper`.
+
+    .. method:: read_bytes()
+
+       Read contents of self as bytes.
```
Comment (hugovk) [inline Doc/library/importlib.rst]:
```suggestion
       Read contents of ``self`` as bytes.
```

File: Doc/library/importlib.rst [hunk 10]
```diff
@@ -645,6 +645,160 @@ ABC hierarchy::
...
+
+    .. method:: read_text(encoding=None)
+
+       Read contents of self as text.
```
Comment (hugovk) [inline Doc/library/importlib.rst]:
```suggestion
       Read contents of ``self`` as text.
```

**Review by serhiy-storchaka — COMMENTED** [review]
Concur with @graingert that separate `versionadded` directive for classes are not needed if they duplicate the module level `versionadded`.

I think that rather than duplicating documentation in `importlib.rst` (it can became de-synchronized), it would be enough to add minimal class declarations followed by references to new classes.

**Comment by hugovk** [comment]
@graingert Would be great to get this ready for 3.12.0 final in a couple of weeks. Thanks!

**Review by jaraco — APPROVED** [review]

**Review by hugovk — APPROVED** [review]
Thank you!

**Comment by hugovk** [comment]
> the versionadded entries are incorrect on the 3.11 doc pages https://docs.python.org/3.11/library/importlib.resources.abc.html#importlib.resources.abc.TraversableResources

Do we need a separate PR to fix this bit in 3.11?

**Comment by graingert** [comment]
@hugovk yeah I think so

**Comment by hugovk** [comment]
@graingert Would you like to do the honours? :)

**Pull request merged**